examples/vision/adamatch.py

```diff
@@ -0,0 +1,453 @@
+weak augmentation, we will use horizontal flipping and random cropping.
+"""
+
+# Initialize `RandAugment` object with 2 layers of
+# augmentation transforms and strength of 5.
+augmenter = RandAugment(num_layers=2, magnitude=5)
+
+
+def weak_augment(image, source=True):
+    if image.dtype != tf.float32:
+        image = tf.cast(image, tf.float32)
+
+    # MNIST images are grayscale, this is why we first convert them to
+    # RGB images.
+    if source:
+        image = tf.image.resize_with_pad(image, RESIZE_TO, RESIZE_TO)
+        image = tf.tile(image, [1, 1, 3])
+    image = tf.image.random_flip_left_right(image)
+    image = tf.image.random_crop(image, (RESIZE_TO, RESIZE_TO, 3))
+    return image
+
+
+def strong_augment(image, source=True):
+    if image.dtype != tf.float32:
+        image = tf.cast(image, tf.float32)
+
+    if source:
+        image = tf.image.resize_with_pad(image, RESIZE_TO, RESIZE_TO)
+        image = tf.tile(image, [1, 1, 3])
+    image = augmenter.distort(image)
+    return image
+
+
+"""
+## Data loader utilities
+"""
+
+
+def create_individual_ds(ds, aug_func, source=True):
+    if source:
+        batch_size = SOURCE_BATCH_SIZE
+    else:
+        # During training 3x more target unlabeled samples are shown
+        # to the model in AdaMatch (Section 3.2 of the paper).
+        batch_size = TARGET_BATCH_SIZE
+    ds = ds.shuffle(batch_size * 10, seed=42)
+
+    if source:
+        ds = ds.map(lambda x, y: (aug_func(x), y), num_parallel_calls=AUTO)
+    else:
+        ds = ds.map(lambda x, y: (aug_func(x, False), y), num_parallel_calls=AUTO)
+
+    ds = ds.batch(batch_size).prefetch(AUTO)
+    return ds
+
+
+"""
+`_w` and `_s` suffixes denote weak and strong respectively.
+"""
+
+source_ds = tf.data.Dataset.from_tensor_slices((mnist_x_train, mnist_y_train))
+source_ds_w = create_individual_ds(source_ds, weak_augment)
+source_ds_s = create_individual_ds(source_ds, strong_augment)
+final_source_ds = tf.data.Dataset.zip((source_ds_w, source_ds_s))
+
+target_ds_w = create_individual_ds(svhn_train, weak_augment, source=False)
+target_ds_s = create_individual_ds(svhn_train, strong_augment, source=False)
+final_target_ds = tf.data.Dataset.zip((target_ds_w, target_ds_s))
+
+"""
+Here's how the image batches look like:
+
+![](https://i.ibb.co/FDVkbWL/Data-Viz.png)
+"""
+
+"""
+## Utilities for loss calculation utilities and weight scheduler
+"""
+
+
+def compute_loss_source(source_labels, logits_source_w, logits_source_s):
+    loss_func = keras.losses.CategoricalCrossentropy(from_logits=True)
+    # First compute the losses between original source labels and
+    # predictions made on the weakly and strongly augmented versions
+    # of the same images.
+    w_loss = loss_func(source_labels, logits_source_w)
+    s_loss = loss_func(source_labels, logits_source_s)
+    return w_loss + s_loss
+
+
+def compute_loss_target(target_pseudo_labels_w, logits_target_s, mask):
+    loss_func = keras.losses.categorical_crossentropy
```

Comment:
Why is this one not from logits if the other one is? This is a bit confusing

Reply:
I am using `from_logits=True` here:

```
target_loss = loss_func(target_pseudo_labels_w, logits_target_s, from_logits=True)
```

Since the reduction of this loss function is a bit more involved I used the `categorical_crossentropy` function rather than the class.

Reply:
Then you should do `loss_func = keras.losses.CategoricalCrossentropy(from_logits=True, reduction='none')`
then `target_loss  = loss_func(target_pseudo_labels_w, logits_target_s)`
to be consistent with the source loss